fix: do not install bindep artifacts by default

rvolosatovs opened the pull request:
## Motivation
<!--
Thank you for your contribution! Please add a brief description below about the change and what is the motivation
behind it.
-->

1. Existing binary installation logic does not work with bindeps feature, since the binaries in the log are not necessarily the build outputs user expected to build, but rather just a dependency build. (I was previously resorting to parsing the list of binaries in the workspace and installing from that https://github.com/rvolosatovs/nixify/blob/3df7d15a90df4cf9ec57c6329b71f74eec4d47ac/lib/rust/mkAttrs.nix#L158-L176 https://github.com/rvolosatovs/nixify/blob/3df7d15a90df4cf9ec57c6329b71f74eec4d47ac/lib/rust/crateBins.nix#L21-L116)
2. ~It's not possible to build `dylib` artifacts, which may be useful in various contexts~ Done in #340
- ~Add support for installing crates utilizing `bindeps` feature by only installing binaries of workspace packages~
- ~Add support for installing `dylib` crate types~

## Checklist
<!--
Note: this list does not have to be complete to submit a contribution!
Fill out what you can and feel free to ask for help with anything
-->
- [x] added tests to verify new behavior
- [ ] added an example template or updated an existing one
- [ ] updated `docs/API.md` (or general documentation) with changes
- [x] updated `CHANGELOG.md`

ipetkov commented on lib/setupHooks/installFromCargoBuildLogHook.sh:
```diff
@@ -13,12 +13,18 @@ function installFromCargoBuildLog() (
 
   local select_non_test='select(.reason == "compiler-artifact" and .profile.test == false)'
-  local select_bins="${select_non_test} | .executable | select(.!= null)"
+  local select_non_dep='select((contains("/release/deps/") or contains("/debug/deps/"))| not)'
```
Should this check be shortened to just `select(contains("/deps/") | not)`? `release` and `dev` aren't the only profiles and additional custom profiles could trip this up

(Also, is there any other metadata that cargo emits in the json output to denote this as a build dep or is the path the only way to check for it?)

rvolosatovs replied:
I did not manage to find anything else than that, which is why I added the profile, just to make sure we're not grabbing something we should not (or we're in some weird edge case with e.g. nix store path set to `/deps` or something like that)

I'll remove the profile

ipetkov commented on lib/setupHooks/installFromCargoBuildLogHook.sh:
```diff
@@ -32,12 +38,11 @@ function installFromCargoBuildLog() (
-      echo "${logs}" | jq -r "${select_member_libs}" | installArtifacts "${dest}/lib"
+      local select_member="select(.package_id == ${ws_member})"
+      echo "${logs}" | jq -r "${select_member} | ${select_lib_files}" | installArtifacts "${dest}/lib"
+      echo "${logs}" | jq -r "${select_member} | ${select_bins}" | installArtifacts "${dest}/bin"
```
I think doing things this way will attempt to install _all binaries_ for a particular member, but that doesn't necessarily mean we have built them all. Consider if the derivation is built with `cargo build -p myCrate --bin foo` but the crate specifies `foo` and `bar` binaries. Would this check also attempt to install `bar`?

Perhaps we should be installing all binaries that were built _and are also not considered dependencies_?

ipetkov replied:
Yeah it would definitely be easier to manage (and test!) with a Rust program. We already have `crane-utils` in the repo (currently only used for vendoring git-dependencies). Feel free to add a new binary in there which would process a cargo installation log!

rvolosatovs replied:
Taking another look, this is actually already using `'"${select_non_dep}"` above, doesn't that cover the issue here?

ipetkov replied:
Looking back at my original comment, I think it might have come out of me misunderstanding the implementation so I'll mark this as resolved!

ipetkov commented on lib/setupHooks/installFromCargoBuildLogHook.sh:
```diff
@@ -13,12 +13,18 @@ function installFromCargoBuildLog() (
+    | select(.target.kind
+        | contains(["cdylib"])
+        or contains(["dylib"])
+        or contains(["lib"])
```
Question: why do we want to install `lib` by default? According to [the reference](https://doc.rust-lang.org/reference/linkage.html) `lib` means 
> This is an ambiguous concept as to what exactly is produced because a library can manifest itself in several forms. The purpose of this generic lib option is to generate the "compiler recommended" style of library. The output library will always be usable by rustc, but the actual type of library may change from time-to-time.

I'm not entirely sure what this means or how it would be usable (but if it's something that can be used by other builds/linkers I don't mind including it; I'd just like to see an example to understand things better!)

rvolosatovs replied:
I'd guess for some niche, very low-level use cases where the consumers of `crane` would want to e.g. build custom `cargoArtifacts`.
Happy to remove it

ipetkov replied:
Yeah I'd like to keep things as simple as possible for now, if someone has a use case which needs it we can revisit

ipetkov commented on lib/setupHooks/installFromCargoBuildLogHook.sh:
```diff
@@ -13,12 +13,18 @@ function installFromCargoBuildLog() (
+        | contains(["cdylib"])
+        or contains(["dylib"])
+        or contains(["lib"])
+        or contains(["rlib"])
```
Question: why do we want to install `rlibs` at all? These are intermediate artifacts used by the compiler and are generally (to my knowledge) not very useful outside of that. Plus the installed cargo artifacts dir should already include these (and in fact installing them all would largely result in installing the contents of cargo's `target` directory IIUC). I'd imagine if someone really needs those artifacts they should probably look for them there and not inside of `$out/lib`

rvolosatovs replied:
SGTM

Pegasust commented:
Would love to have this merged (I need `dylib`). Is there anything blocking this that I may take on to help?

rvolosatovs commented:
> Would love to have this merged (I need `dylib`). Is there anything blocking this that I may take on to help?

I'm just really swamped with work - rebased and addressed the feedback, don't have time to run the checks locally, so will (ab-)use CI for that, but I will pull in the newest version in my downstream project to test this.

I think all @ipetkov comments are addressed. @ipetkov if you agree, I'd propose filing an issue about a Rust program parsing the log. My intention was to port/adapt https://github.com/wasmCloud/wasmCloud/blob/934d70121e768eb33921bdaf5107881ba91ea963/tests/actors/build.rs#L13-L228, I was in fact, writing this code both thinking about how crane handles this and planning to contribute

ipetkov commented:
@rvolosatovs I'll try to take a closer look sometime this week, I've cached the details out of working memory and need to remember what my original comments meant or if they're still relevant :sweat_smile: (will play around with cargo's log format and see what data it gives us)

In the mean time I'm going to split out and land the `dylib` portion of the MR separately. I think that part is pretty uncontroversial and we can definitely land it as is, then focus on the bindeps portion more closely!

ipetkov commented:
@Pegasust feel free to track #340 for `dylib` support!

ipetkov commented on lib/setupHooks/installFromCargoBuildLogHook.sh:
```diff
@@ -13,12 +13,17 @@ function installFromCargoBuildLog() (
 
   local select_non_test='select(.reason == "compiler-artifact" and .profile.test == false)'
-  local select_bins="${select_non_test} | .executable | select(.!= null)"
+  local select_non_dep='select(contains("/deps/")| not)'
```
Played around with the bindeps feature myself and I found out that (fortunately) cargo will emit different log entries if a workspace target happens to _also_ be used as a bindep for another member. This means that we can always unconditionally ignore anything in `.../deps/artifact/...`, but we'll still install the actual targets if they happened to be part of the `cargo build` invocation, so modulo some comments here I'm happy to merge this in!
